design/plankton/CommunicationLayerDesign.md

```diff
@@ -36,13 +36,161 @@ The IBEX team at ISIS, for example, have expressed interest in using Plankton. N
+
+
+
+#### Environment & Timing Loop
+
+The design separates the timing loop from the `Adapter`. The while loop is built into an `Environment`, of which there is just one per simulation. We would like to be able to make updates to the `Environment` without having ot restart the simulation or renew the adapters. One possible implementation of the `Environment` might look like this:
+
+```python
+
+class Environment(object):
+
+    def __init__(self, adapter):
+        self._adapter = adapter # adapter wraps device (see above)
+        self._count = 0
+        self._timer = 0.0
+        self._delta = 0.0
+        self._user_delta = None
+        
+    def _delta(self):
+        if self._user_delta:
+            return self._user_delta
+        else:
+            return (datetime.now() - start).total_seconds()
+            
+    def set_user_delta(self, delta):
+        self._user_delta = delta
+
+    def run(self):
+
+        while True:
+            start = datetime.now()
+
+            
+            self._adapter.process(delta) # stream adapters would call asyncore.loop(0.1, count=1) as part of this
+
+            
+            count += 1
+            timer += self._delta()
+            self._delta = self._delta()
+            # ....
+```
+
+#### Environment Updates
+
+The separation of the runnable parts of the simulation (now in the `Environment`) from the `Device` and the `Adapter` parts should make updates which affect only the `Environment` easier.
+
+Here's one possibility I've been considering:
+
+`EnvironmentAdapter` sees `Environment` 
+
+```python 
+    class Environment(object):
+
+    def __init__(self, adapter, environ_adapter):
+        self._environ_adapter = environ_adapter
+        self._environ_adapter.set_adaptee(self)
+        self._adapter = adapter
+
+    def run(self):
+
+        while True:
+            # Everything still driven from a central processing loop.
+            self._environment_adapter.process()
+            
+            self._adapter.process(delta)
+
+```
+
+```python
+   class StreamEnvironmentAdapter(object):
+
+    def set_adaptee(self, environment):
+        self._environment = environment
+        
+    
+    def process(self):
+        asyncore.loop(0.1, count=1)
+        # ...
+        self._environment.set_user_delta(user_value)
+        
+```
+
+After discussion with Michael Wedel, I am now strongly leaning towards providing a simple single protocol for the `Environment Control`. While `Adapter` flexibility and variation is vital in order to support different device types. the `Environment` is internal to plankton and we should publish and control a single mechanism for this. Michael Wedel has suggested JSON over zeromq.
```

Comment:
@MichaelWedel 

I took a look at the solution in live-data-prototype, and I really like the idea. I think you're right, this is a better solution than a telnet interface. If we ever really need a telnet interface for something, we could build it on top of it.

In addition to the things Owen mentioned, I'd consider how this might be used from a Python script (like a unit test that wants to control the simulation). It would be good to be able to just import a module and use it as a client directly. The `control.py` (or whatever we call the command-line interface) could then use that as well / wrap around it.

That's already almost what's there, it's just that:

- It's a bit awkward that you need two imports / instances (parameter_control_client + general_parameter_controller)
- Some of the controller's methods print to stdout rather than returning information